docs/source/app-dev/command-deduplication.rst

```diff
@@ -0,0 +1,339 @@
+However, the application typically cannot distinguish a lost submission from a Daml ledger that is just slow in processing the submissions.
+So the application should just resubmit the command until it is executed against the ledger, and rely on command deduplication to reject all duplicate submissions that do make it to the ledger.
+
+Some ledger changes can be executed at most once anyway.
+No command deduplication is needed for them.
+For example, if the submitted command exercises a consuming choice on a given contract ID, this command can be accepted at most once because every contract can be archived at most once.
+All duplicate submissions are rejected with :ref:`CONTRACT_NOT_ACTIVE <error_code_CONTRACT_NOT_ACTIVE>`.
+
+In contrast, a :ref:`Create command <com.daml.ledger.api.v1.CreateCommand>` would create a fresh contract instance of the given :ref:`template <com.daml.ledger.api.v1.CreateCommand.template_id>` for each submission that reaches the ledger (unless other constraints such as the :ref:`template preconditions <daml-ref-preconditions>` or contract key uniqueness are violated).
+Similarly, an :ref:`Exercise command <com.daml.ledger.api.v1.ExerciseCommand>` on a non-consuming choice or an :ref:`Exercise-By-Key command <com.daml.ledger.api.v1.ExercisebyKeyCommand>` may be executed multiple times if submitted multiple times.
+With command deduplication, applications can ensure such intended ledger changes are executed only once, even if the application resubmits within the deduplication period, say because it considers the earlier submissions to be lost or forgot during a crash that it had already submitted the command.
+
+As a typical example, the application wants to create exactly one contract instance of template ``T`` via the corresponding :ref:`Create command <com.daml.ledger.api.v1.CreateCommand>`.
+We assume that the ledger is correctly set up so that the Create command should be accepted, i.e., Daml packages are uploaded and vetted, the parties are allocated, and the application has a valid access token for the submission.
+
+
+Known processing time bounds
+============================
+
+For now, we assume that the application knows a bound ``B`` on the processing time and forward clock drifts (against the application’s clock) in the Daml ledger.
+If processing measured across all retries takes longer than the bound ``B``, the ledger change may take effect several times.
+Under this assumption, the following strategy works for applications that use the :ref:`Command Service <command-service>` or the :ref:`Command Submission <command-submission-service>` and :ref:`Command Completion Service <command-completion-service>`.
+
+.. _dedup-bounded-step-command-id:
+
+#. Choose a fresh command ID for the ledger change and the ``actAs`` parties, which (together with the application ID) determine the change ID.
+   Remember the command ID across application crashes.
```

Comment:
If the application were to use a random UUID for the command ID to identify the ledger change, then the application must persist that it has picked a given UUID for some fixed intended ledger change so that if the application crashes and restarts, it will use the same UUID again for resubmissions and not pick a fresh UUID. That's why we recommend to derive the command ID deterministicly.

Reply:
Oh I totaly misunderstood (my brain picked `crashes` as a verb for some reason). Perhaps expanding a bit can help, e.g. `Persist the command ID so that, if the application crashes and restarts, it will use the same command ID for resubmissions`.

Reply:
Expanded.